### ISSUE TITLE
Speed up by TensorRT int8 mode.

### ISSUE THREAD

**Issue by RockyXu66** [TeCSAR-UNCC/EfficientHRNet#4]
Hello, thanks for your work. I want to deploy your amazing model in my jetson xavier nx. Your paper only mentioned using TensorRT on xavier nx, achieving 23fps performance. I am curious which mode you are using, fp32, fp16 or int8. Did you try to quantize the model in Int8 mode for further speedup?

**Comment by ghost** [comment]
We used FP16. For our purposes the performance was more than enough, and we wanted to maintain an accuracy as close to our reported results as possible. We did not test with Int 8, though I would be interested in your results if you do.

**Comment by RockyXu66** [comment]
@CGNefarious Sure, I will put the result in here after my attempt. Btw, could you tell me which version of the TensorRT you are using? There might be some layers unsupported for specific version of TensorRT when calibrating the model.

**Comment by ghost** [comment]
Sorry for the late response. The issue is likely with the upsample layers. You need to define static sizes instead of using a scaling factor. Also play around with the opset version of ONNX. I used 10, but it doesn't take much to change the opset version to see what works. TensorRT version 7.1.3 is what I used. Hope that helps.

**Comment by oscarthorn** [comment]
Do you happen to have the code/procedure for running the model with tensorrt available? I curious to try it out, thinking of doing an implementation of your model for a project I'm working on. Really impressive work by the way, especially the speed is really impressive.

**Comment by RockyXu66** [comment]
@CGNefarious 
Hi, I have some problems when deploying your model under the machine of Jetson Xavier NX. First, I converted the H0 torch model to tensorrt engine with fp16 and static input size 512x416. The inference time is around 40ms. However, when I counted total elapsed time and every step of elapsed time, I found that the machine spends a lot of time on post processing. The "upsampling output" process takes around 77ms. The "Non-maximum-suppression" process takes around 91ms. The "tag matching" process takes around 15ms. 

Your paper mentioned that you achieve 23fps under Xavier NX. I'm wondering what is your static input size for H0 tensorrt engine. Does the 23fps performance include preprocessing and post-processing or just the engine inference process? If 23fps includes the whole pipeline, I am guessing maybe you converted the whole python code into C++ and use the cuda kernel for faster post-processing?

**Comment by oscarthorn** [comment]
Any update on this? One of the major contributions is the speed of your model so it would be interesting to test. Or to get clarification that the fps only includes the network and not the post-processing to turn the output into linked keypoints.

**Comment by ghost** [comment]
To clarify, the latency and resulting FPS in the paper is specifically for the model inferencing time. To achieve high throughput on the NX I would suggest reimplementing the pre and post processing in C++ and optimizing it for speed. The pytorch version of the code provided was used for training and is very slow for inference purposes.

Our lab currently developing this C++ implementation for another project (I have been moved to work on something else) so if that project is ever published I will link the code here. Our manpower has greatly diminished in the past year due to COVID and other issues, so unfortunately I can't give you an estimate on when that will be available. But from the work I did on the conversion I can tell you it is a non-trivial engineering task.

**Comment by ghost** [comment]
Our input sizes for each model are listed in the tables on the README and in our paper. The are square RGB images, so 512 will become 512x512x3. We did briefly explore training for other aspect ratios, but the initial results were not promising so we did not pursue that avenue.

**Comment by oscarthorn** [comment]
Thanks for the clarification! I'll look at writing move own pipeline then maybe.

Anyway thanks again for the great work!
